README
Les scripts de mysrava sont livrés tels quels.
Pour fonctionner ils nécessitent plusieurs actions de paramétrage.

1 - Créer une "Application API" dans votre compte Strava (http://www.strava.com)
2 - A la racine d'installation de mystrava
			Renommer le fichier strava-config-sample.inc.php en strava-config.inc.php
3 - Reporter les 3 paramètres de votre application Strava dans le fichier strava-config.inc.php
4 - Tester la connexion : http://site/mystrava/public_strava.php
		Vous devriez voir apparaître le résumé des activités "public"
5 - Tester la connexion : http://site/mystrava/private_strava.html
		Vous devriez voir apparaître le résumé de l'ensemble de vos activités depuis 
		le 1 janvier de l'année en cours dans la limite des 200 dernières activités.

En cas de problème il va falloir procéder à une analyse particulière.


ChangeLog
mystrava ChangeLog
==================

2017-01-28 17:55 laurent

	* First version of package. Raw display of data. - release 0.0.1


sources/libstrava.php
<?php


// Classe permettant la réalisation de requetes vers le site Strava
class mystrava {

	//**** PUBLIC ****
	// Constructeur
	public function __construct() {
		$this->url = "";/*{{{*/
		$this->tocken = "";
		$this->result = "";
		$this->client_id = "";
		$this->redirect_uri = "";
		$this->response_type = "";
		$this->result = "";
		$this->options = "";
		$this->per_page = "";
		$this->page = "";
		$this->before = "";
		$this->after = "";
		$this->state = "";
		$this->approval_prompt = "";
		$this->client_secret = "";
		$this->access_token = "";

		$this->testing = 0;
		$this->debug = 0;
	}/*}}}*/

	// Destructeur
	public function __destruct() {
	}

	// Getter et Setter
/*{{{*/
	// Getter testing - Variable utilisée pour afficher au lien d'exécuter
	public function gettesting() {
		return $this->testing;
	}

	// Setter testing - Variable utilisée pour afficher au lien d'exécuter
	public function settesting($tst) {
		$this->testing = $tst;
	}

	// Getter testing - Variable utilisée pour afficher le résultat des commandes
	public function getdebug() {
		return $this->debug;
	}

	// Setter debug - Variable utilisée pour afficher le résultat des commandes
	public function setdebug($dbg) {
		$this->debug = $dbg;
	}

	// Getter result
	public function getresult() {
		return $this->result;
	}

	// Setter result
	public function setresult($res) {
		$this->result = $res;
	}

	// Getter token
	public function gettoken() {
		return $this->token;
	}
	// Setter token
	public function settoken($tok) {
		$this->token = $tok;
	}

	// Getter url
	public function geturl() {
		return $this->url;
	}
	// Setter url
	public function seturl($ur) {
		$this->url = $ur;
	}

	// Getter client_id
	public function getclient_id() {
		return $this->client_id;
	}
	// Setter client_id
	public function setclient_id($clt_id) {
		$this->client_id = $clt_id;
	}

	// Getter redirect_uri
	public function getredirect_uri() {
		return $this->redirect_uri;
	}
	// Setter redirect_uri
	public function setredirect_uri($uri) {
		$this->redirect_uri = $uri;
	}

	// Getter response_type
	public function getresponse_type() {
		return $this->response_type;
	}
	// Setter response_type
	public function setresponse_type($resp_typ) {
		$this->response_type = $resp_typ;
	}

	// Getter approval_prompt
	public function getapproval_prompt() {
		return $this->approval_prompt;
	}
	// Setter approval_prompt
	public function setapproval_prompt($prt) {
		$this->approval_prompt = $prt;
	}

	// Getter per_page
	public function getper_page() {
		return $this->per_page;
	}
	// Setter per_page
	public function setper_page($pp) {
		$this->per_page = $pp;
	}

	// Getter page
	public function getpage() {
		return $this->page;
	}
	// Setter page
	public function setpage($p) {
		$this->page = $p;
	}

	// Getter after
	public function getafter() {
		return $this->after;
	}
	// Setter after
	public function setafter($aft) {
		$this->after = $aft;
	}

	// Getter before
	public function getbefore() {
		return $this->before;
	}
	// Setter before
	public function setbefore($befo) {
		$this->before = $befo;
	}

	// Getter state
	public function getstate() {
		return $this->state;
	}
	// Setter state
	public function setstate($st) {
		$this->state = $st;
	}

	// Getter scope
	public function getscope() {
		return $this->scope;
	}
	// Setter scope
	public function setscope($sco) {
		$this->scope = $sco;
	}

	// Getter client_secret
	public function getclient_secret() {
		return $this->client_secret;
	}
	// Setter client_secret
	public function setclient_secret($secret) {
		$this->client_secret = $secret;
	}

	// Getter access_token
	public function getaccess_token() {
		return $this->access_token;
	}
	// Setter access_token
	public function setaccess_token($atk) {
		$this->access_token = $atk;
	}



/*}}}*/

	// Fonction permettant de réaliser l'appel au site Strava
 	public function run_request() {/*{{{*/
		if($this->url != "") {
			$this->build_options();

			if($this->testing != 0) {
				echo "Test URL appelée : ";
				echo $this->url.$this->options;
				echo "<br />";
			}
			else {
				$this->result = file_get_contents($this->url."".$this->options);
				if($this->debug != 0) {
					echo "Debug URL appelée : ";
					echo $this->url.$this->options;
					echo "<br />";
					print_r($this->result);
					echo "<br />";
				}
			}
		}
		else {
			$this->result = "Développeur, vous devez initialiser au minimum l'URL avant de faire appel à l'objet...";
			if($this->debug != 0 or $this->testing != 0) {
				echo "<br />".$this->result."<br />";
			}
		}
	}/*}}}*/

	//**** PROTECTED ****


	//**** PRIVATE ****

	// Fonction pour obtenir access_token après authentification et autorisation de l'application
	public function get_access_token($secret, $code) { /*{{{*/
		$this->setclient_secret($secret);
		$params = array();

		if($this->client_id != "") {
			// Les variables sont correctement positionnées et on peut procéder à l'éxecution
			$params["client_id"] = $this->getclient_id();
			$params["client_secret"] = $this->getclient_secret();
			$params["code"] = "$code";
			$query = http_build_query ($params);	// Build Http query using params
			// Create Http context details
			$contextData = array (
				'method' => 'POST',
				'header' => "Connection: close\r\n".
				"Content-Length: ".strlen($query)."\r\n",
				'content'=> $query );
			// Create context resource for our request
			$context = stream_context_create (array ( 'http' => $contextData ));
			// Read page rendered as result of your POST request
			$result =  file_get_contents ( $this->url, false, $context);
			$json = json_decode($result);
			$this->access_token = $json->access_token;
			if($this->testing!=0 and $this->debug!=0) {
				print_r($json);
				echo "Token : ".$this->access_token;
			}
		}
		else {
			// Ces variables doivent impérativement être définies
			// Peut-être sera-t-il nécessaire de définir des valeurs de retour...
			return false;
		}
	}
	/*}}}*/

	// Fonction d'ajout des caractères spéciaux pour la requête HTTP
	private function add_speccar() { /*{{{*/
		if($this->options != "")
			$this->options = $this->options."&";
		else
			$this->options = $this->options."?";
	} /*}}}*/

	// Fonction permettant de construire les options de la requête Strava
	private function build_options() { /*{{{*/
	
		if($this->token!="" or $this->access_token!="") {
			$this->add_speccar();
			if($this->debug!=0) {
				echo "Token : ".$this->token." Access_token : ".$this->access_token."<br />";
			}
			if($this->access_token != "")	// S'il y a identification privée alors je l'utilise
				$this->options = $this->options."access_token=".$this->access_token;
			else	// Sinon utilisation du profil public
				$this->options = $this->options."access_token=".$this->token;
		}
		if($this->client_id !=  ""){
			$this->add_speccar();
			$this->options = $this->options."client_id=$this->client_id";
		}
		if($this->redirect_uri != ""){
			$this->add_speccar();
			$this->options = $this->options."redirect_uri=$this->redirect_uri";
		}
		if($this->response_type != ""){
			$this->add_speccar();
			$this->options = $this->options."response_type=$this->response_type";
		}
		if($this->approval_prompt != ""){
			$this->add_speccar();
			$this->options = $this->options."approval_prompt=$this->approval_prompt";
		}
		if($this->per_page !=  ""){
			$this->add_speccar();
			$this->options = $this->options."per_page=$this->per_page";
		}
		if($this->page !=  "") {
			$this->add_speccar();
			$this->options = $this->options."page=$this->page";
		}
		if($this->after != ""){
			$this->add_speccar();
			$this->options = $this->options."after=$this->after";
		}
		if($this->before!= ""){
			$this->add_speccar();
			$this->options = $this->options."before=$this->before";
		}
		if($this->state!=""){
			$this->add_speccar();
			$this->options = $this->options."state=$this->state";
		}
		if($this->scope!=""){
			$this->add_speccar();
			$this->options = $this->options."scope=$this->scope";
		}

	} /*}}}*/

	// Paramètres optionnels
	private $client_id;
	private $redirect_uri;
	private $response_type;
	private $approval_prompt;
	private $per_page;
	private $after;
	private $before;
	private $client_secret;

	// Paramètres obligatoires
	private $token;
	private $url;
	
	// Résultats
	private $result;
	private $options;
	private $access_token;
	


	// Variables de développement
	private $testing;	// 0 : execution, 1: test (echo)

}


?>


sources/public_strava.php
	<?php

	include_once "strava_fixed/url.inc.php";
	include_once "libstrava.php";
	include_once "strava-config.inc.php";

	//$id_client="1280";

	if(count($_GET) >=1 && $mydebug!=0)
		var_dump($_GET);

	$year = date("Y");
	$premieran = mktime(0,0,0,1,1,$year);

	$obj_public = new mystrava();
	$obj_public->settesting($mytesting);	// Set the object debug mode 
	$obj_public->setdebug($mydebug);

	$obj_public->settoken($public_token);
	$obj_public->seturl($activities_url);
	$obj_public->setper_page(200);
	$obj_public->setafter($premieran);
	$obj_public->run_request();
	$response = $obj_public->getresult();

	$json = json_decode($response);

	$dist_totale = 0;
	$elev_tot = 0;
	$move = 0;
	$altitude_max = 0;
	$commute_nb = 0;
	$sup = array();

	//echo "Nombre d'enregistrements : ".count ($json)."<br />";
	//print_r($json);

		foreach($json as $key1 => $object_act) {
			foreach($object_act as $key2 => $activ) {
				if(! is_object($activ)) {
					if($key2 == "distance") {
						$dist_totale = $dist_totale+$activ;
						$km = round($activ/1000,2);
						
						if($km<90)
							$index="inférieure(s) à 90";
						else if($km>=90 and $km<100)
							$index="entre 90 et 100";
						else if($km>=100 and $km<150)
							$index="entre 100 et 150";
						else if($km>=150 and $km<200)
							$index="entre 150 et 200";
						else if($km>=200)
							$index="inférieure(s) à 200";
						else
							$index="autres";

						if( isset($sup["$index"]) )
							$sup["$index"] = $sup["$index"]+1;
						else
							$sup["$index"] = 1;

						//echo "$km - ".$sup["$index"]." - ".$object_act->start_date."<br />";
					}
					if($key2 == "total_elevation_gain")
						$elev_tot = $elev_tot+$activ;
					if($key2 == "moving_time")
						$move = $move+$activ;
					if($key2 == "elev_high") {
						if($activ > $altitude_max)
							$altitude_max = $activ;
					}
					if($key2 == "commute") {
						if($activ == 1) {
							$commute_nb++;
						}
					}
				}
			}
		}

	$dist_totale = $dist_totale/1000;
	$move = $move/3600;
	echo "<h2>Données publiques de l'année $year</h2>";
	echo "Nombre de sorties : ".count($json)."<br />";
	echo "Distance totale : ".round($dist_totale,2)."km<br />";
	echo "Dénivelé positif : ".$elev_tot."m<br />";
	echo "Temps de déplacement : ".round($move,2)."h<br />";
	echo "Altitude maximum de l'année : ".$altitude_max."m<br />";
	echo "Nombre de commute(s) : ".$commute_nb."</br>";
	arsort($sup);
	$keys = array_keys($sup);
	foreach ($keys as $dist_key) {
		$kvalue = $sup["$dist_key"];
		$percent = round($sup["$dist_key"]*100/count($json));
		echo "Nombre de sorties ".$dist_key."km : ".$sup["$dist_key"]." ($percent % des sorties)</br>";
	}

	?>


sources/private_strava.php
<?php
include_once "strava_fixed/url.inc.php";
include_once "libstrava.php";
include_once "strava-config.inc.php";

if( $_GET['code'] && ! $_COOKIE["mystrava_token"]) {
	$code = $_GET['code'];	// Get authorized code from HTTP response
}
else {
	$code = null;
}

$obj = new mystrava;

// Debug section
$obj->settesting($mydebug);
$obj->setdebug($mytesting);

$obj->setclient_id($id_client);


if(! $_COOKIE["mystrava_token"]) {
	if( $code == null ) {
		$url="https://www.strava.com/oauth/authorize?client_id=$id_client&redirect_uri=http://velo.famille-guignard.org/mystrava/private_strava.php&response_type=code&approval_prompt=force&state=mystate&scope=view_private";
		header("Location: $url");
	}
	else {
		$obj->seturl($token_url);
		$obj->get_access_token($secret_client,$code);
		setcookie("mystrava_token",$obj->getaccess_token(),time()+60*60*24*120,"/mystrava/");
	}
}
else {
	$obj->setaccess_token($_COOKIE["mystrava_token"]);
}

	$year = date("Y");
	$date = mktime(0,0,0,1,1,$year);
	$premieran = mktime(0,0,0,1,1,date("Y"));

	$dist_totale = 0;
	$elev_tot = 0;
	$move = 0;
	$altitude_max = 0;
	$commute_nb = 0;
	$lastdate = "";
	$sup = array();


	$obj->seturl($activities_url);
	$obj->setper_page(200);
	$obj->setafter($premieran);
	$obj->run_request();
	$response = $obj->getresult();

	$json = json_decode($response);

		// Analyse des résultats - partie statistique
		foreach($json as $key1 => $object_act) {
			foreach($object_act as $key2 => $activ) {
				if(! is_object($activ)) {
					if($key2 == "distance") {
						$dist_totale = $dist_totale+$activ;
						$km = round($activ/1000,2);
						
						if($km<90)
							$index="inférieure(s) à 90";
						else if($km>=90 and $km<100)
							$index="entre 90 et 100";
						else if($km>=100 and $km<150)
							$index="entre 100 et 150";
						else if($km>=150 and $km<200)
							$index="entre 150 et 200";
						else if($km>=200)
							$index="inférieure(s) à 200";
						else
							$index="autres";

						if( isset($sup["$index"]) )
							$sup["$index"] = $sup["$index"]+1;
						else
							$sup["$index"] = 1;
					}
					if($key2 == "total_elevation_gain")
						$elev_tot = $elev_tot+$activ;
					if($key2 == "moving_time")
						$move = $move+$activ;
					if($key2 == "elev_high") {
						if($activ > $altitude_max)
							$altitude_max = $activ;
					}
					if($key2 == "commute") {
						if($activ == 1) {
							$commute_nb++;
						}
					}
					if($key2 == "start_date") {
						//echo "$activ / ";
						$lastdate = $activ;
					}
				}
			}
		}


	// Exploitation des résultats : affichage
	$dist_totale = $dist_totale/1000;
	$move = $move/3600;
	echo "<h2>Données strava pour l'année $year</h2>";
	echo "Date de la dernière activité traitée : $lastdate<br />";
	echo "Nombre de sorties : ".count($json)."<br />";
	echo "Distance totale : ".round($dist_totale,2)."km<br />";
	echo "Dénivelé positif : ".$elev_tot."m<br />";
	echo "Temps de déplacement : ".round($move,2)."h<br />";
	echo "Altitude maximum de l'année : ".$altitude_max."m<br />";
	echo "Nombre de commute(s) : ".$commute_nb."</br>";
	arsort($sup);
	$keys = array_keys($sup);
	foreach ($keys as $dist_key) {
		$kvalue = $sup["$dist_key"];
		$percent = round($sup["$dist_key"]*100/count($json));
		echo "Nombre de sorties ".$dist_key."km : ".$sup["$dist_key"]." ($percent % des sorties)</br>";
	}
	


?>
